File: python/rafter_cli/commands/notify.py
"""rafter notify — post scan results to Slack/Discord channels."""
from __future__ import annotations

import json
import os
import sys
import urllib.request
from typing import Optional

import typer

from ..utils.api import (
    API_BASE,
    API_TIMEOUT_SHORT,
    EXIT_GENERAL_ERROR,
    EXIT_SCAN_NOT_FOUND,
    EXIT_SUCCESS,
    resolve_key,
)
from ..utils.formatter import fmt, print_stderr

notify_app = typer.Typer(
    name="notify",
    help="Post scan results to Slack or Discord channels via webhooks.",
    invoke_without_command=True,
    no_args_is_help=False,
)


def _resolve_webhook(cli_opt: str | None) -> str:
    """Resolve webhook URL from CLI option, env var, or config."""
    if cli_opt:
        return cli_opt
    env_url = os.getenv("RAFTER_NOTIFY_WEBHOOK")
    if env_url:
        return env_url
    # Try config file
    try:
        from ..core.config_manager import ConfigManager
        config = ConfigManager.load()
        if config.agent.notifications.webhook:
            return config.agent.notifications.webhook
    except Exception:
        pass
    print("No webhook URL provided. Use --webhook or set RAFTER_NOTIFY_WEBHOOK", file=sys.stderr)
    raise typer.Exit(code=EXIT_GENERAL_ERROR)


def _detect_platform(url: str) -> str:
    """Detect webhook platform from URL pattern."""
    if "hooks.slack.com" in url or "slack.com/api" in url:
        return "slack"
    if "discord.com/api/webhooks" in url or "discordapp.com/api/webhooks" in url:
        return "discord"
    return "generic"


def _format_slack_payload(scan_data: dict) -> dict:
    """Format scan results as a Slack message with Block Kit."""
    status = scan_data.get("status", "unknown")
    repo = scan_data.get("repository_name", "unknown")
    scan_id = scan_data.get("scan_id", "")
    findings = scan_data.get("findings", [])
    summary = scan_data.get("summary", {})

    critical = summary.get("critical", 0)
    high = summary.get("high", 0)
    medium = summary.get("medium", 0)
    low = summary.get("low", 0)
    total = critical + high + medium + low

    # Status emoji
    if status == "completed" and total == 0:
        status_icon = ":white_check_mark:"
        status_text = "Clean — no issues found"
    elif status == "completed" and (critical > 0 or high > 0):
        status_icon = ":rotating_light:"
        status_text = f"{total} issue{'s' if total != 1 else ''} found"
    elif status == "completed":
        status_icon = ":warning:"
        status_text = f"{total} issue{'s' if total != 1 else ''} found"
    elif status == "failed":
        status_icon = ":x:"
        status_text = "Scan failed"
    else:
        status_icon = ":hourglass_flowing_sand:"
        status_text = f"Scan {status}"

    blocks: list[dict] = [
        {
            "type": "header",
            "text": {
                "type": "plain_text",
                "text": f"{status_icon} Rafter Security Scan",
            },
        },
        {
            "type": "section",
            "fields": [
                {"type": "mrkdwn", "text": f"*Repository:*\n{repo}"},
                {"type": "mrkdwn", "text": f"*Status:*\n{status_text}"},
            ],
        },
    ]

    if scan_id:
        blocks[1]["fields"].append(
            {"type": "mrkdwn", "text": f"*Scan ID:*\n`{scan_id}`"}
        )

    branch = scan_data.get("branch_name")
    if branch:
        blocks[1]["fields"].append(
            {"type": "mrkdwn", "text": f"*Branch:*\n`{branch}`"}
        )

    # Severity breakdown
    if total > 0:
        severity_parts = []
        if critical:
            severity_parts.append(f":red_circle: Critical: *{critical}*")
        if high:
            severity_parts.append(f":orange_circle: High: *{high}*")
        if medium:
            severity_parts.append(f":large_yellow_circle: Medium: *{medium}*")
        if low:
            severity_parts.append(f":white_circle: Low: *{low}*")

        blocks.append({
            "type": "section",
            "text": {
                "type": "mrkdwn",
                "text": "\n".join(severity_parts),
            },
        })

    # Top findings (max 5)
    if findings:
        finding_lines = []
        for f in findings[:5]:
            sev = f.get("severity", "unknown").upper()
            title = f.get("title", f.get("rule_id", "Unknown"))
            location = f.get("location", f.get("file", ""))
            line = f"• `[{sev}]` {title}"
            if location:
                line += f" — {location}"
            finding_lines.append(line)

        if len(findings) > 5:
            finding_lines.append(f"_... and {len(findings) - 5} more_")

        blocks.append({"type": "divider"})
        blocks.append({
            "type": "section",
            "text": {
                "type": "mrkdwn",
                "text": "*Top Findings:*\n" + "\n".join(finding_lines),
            },
        })

    blocks.append({
        "type": "context",
        "elements": [
            {"type": "mrkdwn", "text": "Posted by *rafter-bot* | <https://rafter.so|rafter.so>"},
        ],
    })

    # Slack needs both text (for notifications) and blocks (for rendering)
    return {
        "text": f"[rafter] {repo}: {status_text}",
        "blocks": blocks,
    }


def _format_discord_payload(scan_data: dict) -> dict:
    """Format scan results as a Discord webhook embed."""
    status = scan_data.get("status", "unknown")
    repo = scan_data.get("repository_name", "unknown")
    scan_id = scan_data.get("scan_id", "")
    findings = scan_data.get("findings", [])
    summary = scan_data.get("summary", {})

    critical = summary.get("critical", 0)
    high = summary.get("high", 0)
    medium = summary.get("medium", 0)
    low = summary.get("low", 0)
    total = critical + high + medium + low

    # Embed color
    if status == "completed" and total == 0:
        color = 0x2ECC71  # green
        status_text = "Clean — no issues found"
    elif status == "completed" and (critical > 0 or high > 0):
        color = 0xE74C3C  # red
        status_text = f"{total} issue{'s' if total != 1 else ''} found"
    elif status == "completed":
        color = 0xF39C12  # orange
        status_text = f"{total} issue{'s' if total != 1 else ''} found"
    elif status == "failed":
        color = 0x95A5A6  # gray
        status_text = "Scan failed"
    else:
        color = 0x3498DB  # blue
        status_text = f"Scan {status}"

    fields: list[dict] = [
        {"name": "Repository", "value": repo, "inline": True},
        {"name": "Status", "value": status_text, "inline": True},
    ]

    if scan_id:
        fields.append({"name": "Scan ID", "value": f"`{scan_id}`", "inline": True})

    branch = scan_data.get("branch_name")
    if branch:
        fields.append({"name": "Branch", "value": f"`{branch}`", "inline": True})

    if total > 0:
        severity_parts = []
        if critical:
            severity_parts.append(f"\U0001f534 Critical: **{critical}**")
        if high:
            severity_parts.append(f"\U0001f7e0 High: **{high}**")
        if medium:
            severity_parts.append(f"\U0001f7e1 Medium: **{medium}**")
        if low:
            severity_parts.append(f"\u26aa Low: **{low}**")
        fields.append({
            "name": "Severity Breakdown",
            "value": "\n".join(severity_parts),
            "inline": False,
        })

    # Top findings
    if findings:
        finding_lines = []
        for f in findings[:5]:
            sev = f.get("severity", "unknown").upper()
            title = f.get("title", f.get("rule_id", "Unknown"))
            location = f.get("location", f.get("file", ""))
            line = f"• `[{sev}]` {title}"
            if location:
                line += f" — {location}"
            finding_lines.append(line)

        if len(findings) > 5:
            finding_lines.append(f"*... and {len(findings) - 5} more*")

        fields.append({
            "name": "Top Findings",
            "value": "\n".join(finding_lines),
            "inline": False,
        })

    embed = {
        "title": "\U0001f6e1\ufe0f Rafter Security Scan",
        "color": color,
        "fields": fields,
        "footer": {"text": "rafter-bot | rafter.so"},
    }

    return {
        "content": f"[rafter] {repo}: {status_text}",
        "embeds": [embed],
    }


def _format_generic_payload(scan_data: dict) -> dict:
    """Format scan results as a plain JSON payload with text field."""
    status = scan_data.get("status", "unknown")
    repo = scan_data.get("repository_name", "unknown")
    summary = scan_data.get("summary", {})
    total = sum(summary.get(k, 0) for k in ("critical", "high", "medium", "low"))

    if status == "completed" and total == 0:
        status_text = "Clean — no issues found"
    elif status == "completed":
        status_text = f"{total} issue{'s' if total != 1 else ''} found"
    else:
        status_text = f"Scan {status}"

    msg = f"[rafter] {repo}: {status_text}"
    return {
        "text": msg,
        "content": msg,
        **scan_data,
    }


def _post_webhook(url: str, payload: dict) -> None:
    """POST JSON payload to a webhook URL."""
    from ..core.audit_logger import validate_webhook_url
    validate_webhook_url(url)

    data = json.dumps(payload).encode("utf-8")
    req = urllib.request.Request(
        url,
        data=data,
        headers={"Content-Type": "application/json"},
        method="POST",
    )
    with urllib.request.urlopen(req, timeout=30) as resp:
        if resp.status >= 400:
            raise RuntimeError(f"Webhook returned HTTP {resp.status}")


def _fetch_scan(scan_id: str, api_key: str) -> dict:
    """Fetch scan results from the backend API."""
    import requests

    headers = {"x-api-key": api_key}
    resp = requests.get(
        f"{API_BASE}/static/scan",
        headers=headers,
        params={"scan_id": scan_id, "format": "json"},
        timeout=API_TIMEOUT_SHORT,
    )
    if resp.status_code == 404:
        print(f"Scan '{scan_id}' not found", file=sys.stderr)
        raise typer.Exit(code=EXIT_SCAN_NOT_FOUND)
    elif resp.status_code != 200:
        print(f"Error: {resp.text}", file=sys.stderr)
        raise typer.Exit(code=EXIT_GENERAL_ERROR)
    return resp.json()


@notify_app.callback()
def notify(
    ctx: typer.Context,
    scan_id: Optional[str] = typer.Argument(None, help="Scan ID to post results for"),
    webhook: Optional[str] = typer.Option(None, "--webhook", "-w", envvar="RAFTER_NOTIFY_WEBHOOK", help="Webhook URL (Slack or Discord)"),
    api_key: Optional[str] = typer.Option(None, "--api-key", "-k", envvar="RAFTER_API_KEY", help="API key for fetching scan results"),
    platform: Optional[str] = typer.Option(None, "--platform", "-p", help="Force platform: slack, discord, or generic"),
    quiet: bool = typer.Option(False, "--quiet", help="Suppress status messages"),
    dry_run: bool = typer.Option(False, "--dry-run", help="Print payload without posting"),
):
    """Post scan results to Slack or Discord channels.

    Provide a scan ID to fetch and post results, or pipe JSON scan data via stdin.

    \b
    Examples:
      rafter notify abc123 --webhook https://hooks.slack.com/services/...
      rafter get abc123 -f json | rafter notify --webhook https://discord.com/api/webhooks/...
      RAFTER_NOTIFY_WEBHOOK=https://hooks.slack.com/... rafter notify abc123
    """
    if ctx.invoked_subcommand is not None:
        return

    webhook_url = _resolve_webhook(webhook)

    # Determine scan data source
    if scan_id:
        key = resolve_key(api_key)
        scan_data = _fetch_scan(scan_id, key)
    elif not sys.stdin.isatty():
        # Read from stdin
        raw = sys.stdin.read()
        try:
            scan_data = json.loads(raw)
        except json.JSONDecodeError:
            print("Error: stdin is not valid JSON", file=sys.stderr)
            raise typer.Exit(code=EXIT_GENERAL_ERROR)
    else:
        print("Error: provide a scan ID or pipe JSON scan data via stdin", file=sys.stderr)
        raise typer.Exit(code=EXIT_GENERAL_ERROR)

    # Detect platform from URL
    detected = platform or _detect_platform(webhook_url)

    # Format payload
    if detected == "slack":
        payload = _format_slack_payload(scan_data)
    elif detected == "discord":
        payload = _format_discord_payload(scan_data)
    else:
        payload = _format_generic_payload(scan_data)

    if dry_run:
        print(json.dumps(payload, indent=2))
        return

    # Post
    if not quiet:
        print_stderr(fmt.info(f"Posting to {detected} webhook..."))

    try:
        _post_webhook(webhook_url, payload)
    except Exception as e:
        print(f"Error posting to webhook: {e}", file=sys.stderr)
        raise typer.Exit(code=EXIT_GENERAL_ERROR)

    if not quiet:
        print_stderr(fmt.success(f"Scan results posted to {detected} channel"))

File: node/src/commands/notify.ts
import { Command } from "commander";
import axios from "axios";
import { API, resolveKey, EXIT_GENERAL_ERROR, EXIT_SCAN_NOT_FOUND } from "../utils/api.js";
import { validateWebhookUrl } from "../core/audit-logger.js";
import { ConfigManager } from "../core/config-manager.js";
import { fmt, isAgentMode } from "../utils/formatter.js";

interface ScanData {
  status?: string;
  repository_name?: string;
  scan_id?: string;
  branch_name?: string;
  findings?: Array<{
    severity?: string;
    title?: string;
    rule_id?: string;
    location?: string;
    file?: string;
  }>;
  summary?: {
    critical?: number;
    high?: number;
    medium?: number;
    low?: number;
  };
  [key: string]: unknown;
}

function resolveWebhook(cliOpt?: string): string {
  if (cliOpt) return cliOpt;
  if (process.env.RAFTER_NOTIFY_WEBHOOK) return process.env.RAFTER_NOTIFY_WEBHOOK;
  // Try config file
  try {
    const configManager = new ConfigManager();
    const config = configManager.load();
    if (config.agent?.notifications?.webhook) {
      return config.agent.notifications.webhook;
    }
  } catch {
    // ignore
  }
  console.error("No webhook URL provided. Use --webhook or set RAFTER_NOTIFY_WEBHOOK");
  process.exit(EXIT_GENERAL_ERROR);
}

function detectPlatform(url: string): string {
  if (url.includes("hooks.slack.com") || url.includes("slack.com/api")) return "slack";
  if (url.includes("discord.com/api/webhooks") || url.includes("discordapp.com/api/webhooks")) return "discord";
  return "generic";
}

function formatSlackPayload(scan: ScanData): Record<string, unknown> {
  const status = scan.status ?? "unknown";
  const repo = scan.repository_name ?? "unknown";
  const scanId = scan.scan_id ?? "";
  const findings = scan.findings ?? [];
  const summary = scan.summary ?? {};
  const critical = summary.critical ?? 0;
  const high = summary.high ?? 0;
  const medium = summary.medium ?? 0;
  const low = summary.low ?? 0;
  const total = critical + high + medium + low;

  let statusIcon: string;
  let statusText: string;
  if (status === "completed" && total === 0) {
    statusIcon = ":white_check_mark:";
    statusText = "Clean — no issues found";
  } else if (status === "completed" && (critical > 0 || high > 0)) {
    statusIcon = ":rotating_light:";
    statusText = `${total} issue${total !== 1 ? "s" : ""} found`;
  } else if (status === "completed") {
    statusIcon = ":warning:";
    statusText = `${total} issue${total !== 1 ? "s" : ""} found`;
  } else if (status === "failed") {
    statusIcon = ":x:";
    statusText = "Scan failed";
  } else {
    statusIcon = ":hourglass_flowing_sand:";
    statusText = `Scan ${status}`;
  }

  const sectionFields: Array<{ type: string; text: string }> = [
    { type: "mrkdwn", text: `*Repository:*\n${repo}` },
    { type: "mrkdwn", text: `*Status:*\n${statusText}` },
  ];
  if (scanId) sectionFields.push({ type: "mrkdwn", text: `*Scan ID:*\n\`${scanId}\`` });
  if (scan.branch_name) sectionFields.push({ type: "mrkdwn", text: `*Branch:*\n\`${scan.branch_name}\`` });

  const blocks: Array<Record<string, unknown>> = [
    { type: "header", text: { type: "plain_text", text: `${statusIcon} Rafter Security Scan` } },
    { type: "section", fields: sectionFields },
  ];

  if (total > 0) {
    const parts: string[] = [];
    if (critical) parts.push(`:red_circle: Critical: *${critical}*`);
    if (high) parts.push(`:orange_circle: High: *${high}*`);
    if (medium) parts.push(`:large_yellow_circle: Medium: *${medium}*`);
    if (low) parts.push(`:white_circle: Low: *${low}*`);
    blocks.push({ type: "section", text: { type: "mrkdwn", text: parts.join("\n") } });
  }

  if (findings.length > 0) {
    const lines = findings.slice(0, 5).map((f) => {
      const sev = (f.severity ?? "unknown").toUpperCase();
      const title = f.title ?? f.rule_id ?? "Unknown";
      const loc = f.location ?? f.file ?? "";
      let line = `• \`[${sev}]\` ${title}`;
      if (loc) line += ` — ${loc}`;
      return line;
    });
    if (findings.length > 5) lines.push(`_... and ${findings.length - 5} more_`);

    blocks.push({ type: "divider" });
    blocks.push({ type: "section", text: { type: "mrkdwn", text: `*Top Findings:*\n${lines.join("\n")}` } });
  }

  blocks.push({
    type: "context",
    elements: [{ type: "mrkdwn", text: "Posted by *rafter-bot* | <https://rafter.so|rafter.so>" }],
  });

  return { text: `[rafter] ${repo}: ${statusText}`, blocks };
}

function formatDiscordPayload(scan: ScanData): Record<string, unknown> {
  const status = scan.status ?? "unknown";
  const repo = scan.repository_name ?? "unknown";
  const scanId = scan.scan_id ?? "";
  const findings = scan.findings ?? [];
  const summary = scan.summary ?? {};
  const critical = summary.critical ?? 0;
  const high = summary.high ?? 0;
  const medium = summary.medium ?? 0;
  const low = summary.low ?? 0;
  const total = critical + high + medium + low;

  let color: number;
  let statusText: string;
  if (status === "completed" && total === 0) {
    color = 0x2ecc71;
    statusText = "Clean — no issues found";
  } else if (status === "completed" && (critical > 0 || high > 0)) {
    color = 0xe74c3c;
    statusText = `${total} issue${total !== 1 ? "s" : ""} found`;
  } else if (status === "completed") {
    color = 0xf39c12;
    statusText = `${total} issue${total !== 1 ? "s" : ""} found`;
  } else if (status === "failed") {
    color = 0x95a5a6;
    statusText = "Scan failed";
  } else {
    color = 0x3498db;
    statusText = `Scan ${status}`;
  }

  const fields: Array<{ name: string; value: string; inline: boolean }> = [
    { name: "Repository", value: repo, inline: true },
    { name: "Status", value: statusText, inline: true },
  ];
  if (scanId) fields.push({ name: "Scan ID", value: `\`${scanId}\``, inline: true });
  if (scan.branch_name) fields.push({ name: "Branch", value: `\`${scan.branch_name}\``, inline: true });

  if (total > 0) {
    const parts: string[] = [];
    if (critical) parts.push(`\u{1f534} Critical: **${critical}**`);
    if (high) parts.push(`\u{1f7e0} High: **${high}**`);
    if (medium) parts.push(`\u{1f7e1} Medium: **${medium}**`);
    if (low) parts.push(`\u26aa Low: **${low}**`);
    fields.push({ name: "Severity Breakdown", value: parts.join("\n"), inline: false });
  }

  if (findings.length > 0) {
    const lines = findings.slice(0, 5).map((f) => {
      const sev = (f.severity ?? "unknown").toUpperCase();
      const title = f.title ?? f.rule_id ?? "Unknown";
      const loc = f.location ?? f.file ?? "";
      let line = `• \`[${sev}]\` ${title}`;
      if (loc) line += ` — ${loc}`;
      return line;
    });
    if (findings.length > 5) lines.push(`*... and ${findings.length - 5} more*`);
    fields.push({ name: "Top Findings", value: lines.join("\n"), inline: false });
  }

  return {
    content: `[rafter] ${repo}: ${statusText}`,
    embeds: [{ title: "\u{1f6e1}\ufe0f Rafter Security Scan", color, fields, footer: { text: "rafter-bot | rafter.so" } }],
  };
}

function formatGenericPayload(scan: ScanData): Record<string, unknown> {
  const status = scan.status ?? "unknown";
  const repo = scan.repository_name ?? "unknown";
  const summary = scan.summary ?? {};
  const total = (summary.critical ?? 0) + (summary.high ?? 0) + (summary.medium ?? 0) + (summary.low ?? 0);

  let statusText: string;
  if (status === "completed" && total === 0) statusText = "Clean — no issues found";
  else if (status === "completed") statusText = `${total} issue${total !== 1 ? "s" : ""} found`;
  else statusText = `Scan ${status}`;

  const msg = `[rafter] ${repo}: ${statusText}`;
  return { text: msg, content: msg, ...scan };
}

export function createNotifyCommand(): Command {
  return new Command("notify")
    .description("Post scan results to Slack or Discord channels via webhooks")
    .argument("[scan_id]", "Scan ID to fetch and post results for")
    .option("-w, --webhook <url>", "Webhook URL (Slack or Discord)")
    .option("-k, --api-key <key>", "API key for fetching scan results")
    .option("-p, --platform <platform>", "Force platform: slack, discord, or generic")
    .option("--quiet", "Suppress status messages")
    .option("--dry-run", "Print payload without posting")
    .action(async (scanId?: string, opts?: Record<string, unknown>) => {
      const webhookUrl = resolveWebhook(opts?.webhook as string | undefined);
      let scanData: ScanData;

      if (scanId) {
        const key = resolveKey(opts?.apiKey as string | undefined);
        try {
          const { data } = await axios.get(`${API}/static/scan`, {
            params: { scan_id: scanId, format: "json" },
            headers: { "x-api-key": key },
          });
          scanData = data;
        } catch (e: any) {
          if (e.response?.status === 404) {
            console.error(`Scan '${scanId}' not found`);
            process.exit(EXIT_SCAN_NOT_FOUND);
          }
          console.error(`Error: ${e.response?.data ?? e.message}`);
          process.exit(EXIT_GENERAL_ERROR);
        }
      } else if (!process.stdin.isTTY) {
        // Read from stdin
        const chunks: Buffer[] = [];
        for await (const chunk of process.stdin) {
          chunks.push(chunk);
        }
        const raw = Buffer.concat(chunks).toString("utf-8");
        try {
          scanData = JSON.parse(raw);
        } catch {
          console.error("Error: stdin is not valid JSON");
          process.exit(EXIT_GENERAL_ERROR);
        }
      } else {
        console.error("Error: provide a scan ID or pipe JSON scan data via stdin");
        process.exit(EXIT_GENERAL_ERROR);
      }

      const detected = (opts?.platform as string) || detectPlatform(webhookUrl);

      let payload: Record<string, unknown>;
      if (detected === "slack") payload = formatSlackPayload(scanData!);
      else if (detected === "discord") payload = formatDiscordPayload(scanData!);
      else payload = formatGenericPayload(scanData!);

      if (opts?.dryRun) {
        console.log(JSON.stringify(payload, null, 2));
        return;
      }

      if (!opts?.quiet) {
        process.stderr.write(fmt.info(`Posting to ${detected} webhook...`) + "\n");
      }

      try {
        await validateWebhookUrl(webhookUrl);
        await fetch(webhookUrl, {
          method: "POST",
          headers: { "Content-Type": "application/json" },
          body: JSON.stringify(payload),
        });
      } catch (e: any) {
        console.error(`Error posting to webhook: ${e.message}`);
        process.exit(EXIT_GENERAL_ERROR);
      }

      if (!opts?.quiet) {
        process.stderr.write(fmt.success(`Scan results posted to ${detected} channel`) + "\n");
      }
    });
}
